src/gridcoin/voting/registry.cpp

```diff
@@ -1060,23 +1076,51 @@ void PollRegistry::DeleteVote(const ContractContext& ctx)
     }
 }
 
+void PollRegistry::DetectReorg()
+EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```

Comment:
This at whole seems redundant/error-prone to me, any re-org that should concern PollRegistry will be signaled to the registry via `PollRegistry::Revert` method from `DisconnectBlocksBatch` with this PR, using that instead feels more solid.

(With a high water canary approach like this, I feel like re-orgs to parallel/ahead chains create edge cases. For example, DetectReorg is called from Add/DeletePoll/Vote whose TXs can be pushed to a higher/equal block height after a re-org to a higher/equal block height; which the high water canary should not detect at all considering the `nBestHeight < m_block_height_hw` check.)

Reply:
You are correct in that the PollRegistry::Revert method is called during block disconnection. That correctly signals that a reorg is in progress. The problem is that it doesn't signal when the reorg is "complete", and we need to know both. I will point out that the Revert for the PollRegistry uses the default contract handler for reversion, which simply calls the reverse action of the original action, which I am taking advantage of here.

void IContractHandler::Revert(const ContractContext& ctx)
{
    if (ctx->m_action == ContractAction::ADD) {
        Delete(ctx);
        return;
    }

    if (ctx->m_action == ContractAction::REMOVE) {
        Add(ctx);
        return;
    }

    error("Unknown contract action ignored: %s", ctx->m_action.ToString());
}


The messiness of this problem I am trying to solve is why @cyrossignol punted and just threw a cs_main around the entire thing, to lock the state of the core during the whole exercise. Problem solved, right? No. We exchange this messiness for a number of other problems, which simply start at people complaining about the frozen GUI. The lock on cs_main for that long a duration causes the nodes response to messages to freeze and the node is then not viewed as a good node to its peers. Since the wallet is not accepting blocks that have been issued from the network during the time cs_main is held, the actual state of the polls during the tally is actually becoming stale. It is certainly correct as of the time of the locking, but not correct as of the actual chain head on the network.

The alternatives aren't perfect either. Simply triggering off of the Revert does not work. There can (and will be) more than one of those possible if the reorg were deep, and furthermore, even when those stop, it isn't the end of the reorg, because another branch has likely been chosen to roll forward to the new head (tip). You are also correct that there is no guarantee that voting transactions resurrected in the mempool (which happens in addition to the revert trigger), will appear on the same block numbers as before and could possibly be at a higher block than the original watermark. I am not sure we care.

In general there is a very high probability that the same transactions will be in the new fork on the reorg, because most of these will cross forks. As long as the output used for the contract send itself is still valid (i.e. wasn't removed due to a stake being removed), the vote will cross the fork and be bound at approximately the same temporal location. Even if it doesn't, so what!

The pattern I am implementing is to detect whether there is a reorg that occurs. If so, abort the tally and hold and poll for the reorg to be complete. Complete is practically defined as the same height as the high water mark. If you don't like that, we could use chain trust, which is what the reorg code in main actually uses to choose forks anyway. Regardless, the idea is to pick a reasonably practical point to RESTART the tally.

The tally will be restarted. Not keeping cs_main locked during this allows each poll when it is tallied to actually use the set of transactions that are valid at the time the tally for that poll was started rather than when the entire refresh loop was started. This is not truly "up to date", but is much closer. If a transaction that had a vote didn't make it back into the chain, then so what. It wasn't supposed to be counted anyway unless it is relinked. If another refresh is done at a later time, it will be there.

We don't really have a choice here except to use a high watermark approach. The choice is height or trust.

Reply:
See the g_reorg_in_progress commit.